src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -1701,16 +1707,47 @@ private IEnumerable<PSTypeName> InferTypesFrom(MemberExpressionAst memberExpress
 
         private static void AddTypesFromMethodCacheEntry(
             DotNetAdapter.MethodCacheEntry methodCacheEntry,
+            IList<ITypeName> genericTypeArguments,
             List<PSTypeName> result,
             bool isInvokeMemberExpressionAst)
         {
             if (isInvokeMemberExpressionAst)
             {
+                Type[] resolvedTypeArguments = null;
+                if (genericTypeArguments is not null)
+                {
+                    resolvedTypeArguments = new Type[genericTypeArguments.Count];
+                    for (int i = 0; i < genericTypeArguments.Count; i++)
+                    {
+                        Type resolvedType = genericTypeArguments[i].GetReflectionType();
+                        if (resolvedType is null)
+                        {
+                            return;
+                        }
+
+                        resolvedTypeArguments[i] = resolvedType;
+                    }
+                }
+
                 foreach (var method in methodCacheEntry.methodInformationStructures)
                 {
-                    if (method.method is MethodInfo methodInfo && !methodInfo.ReturnType.ContainsGenericParameters)
+                    if (method.method is MethodInfo methodInfo)
                     {
-                        result.Add(new PSTypeName(methodInfo.ReturnType));
+                        if (!methodInfo.ReturnType.ContainsGenericParameters)
```

Comment:
Can we join the two `if`-s?

Reply:
Sure it's possible but I don't think it makes sense to do so. I could add `method.method is MethodInfo methodInfo &&` in front of both if/else if statements but that would make it harder to read (IMO) and from my understanding of "is" perform worse.

Reply:
Usually we prefer to reduce nesting.

Reply:
Sure but I just don't think it makes sense here. Compare these 2:

Current:

```
if (method.method is MethodInfo methodInfo)
{
    if (!methodInfo.ReturnType.ContainsGenericParameters)
    {
        result.Add(new PSTypeName(methodInfo.ReturnType));
    }
    else if (genericTypeArguments is not null)
    {
        try
        {
            result.Add(new PSTypeName(methodInfo.MakeGenericMethod(resolvedTypeArguments).ReturnType));
        }
        catch // If we can't build the generic method then just skip it to retain other completion results.
        {
            return;
        }
    }
}
```

Suggested:

```
if (method.method is MethodInfo methodInfo && !methodInfo.ReturnType.ContainsGenericParameters)
{
    result.Add(new PSTypeName(methodInfo.ReturnType));
}
else if (method.method is MethodInfo methodInfo2 && genericTypeArguments is not null)
{
    try
    {
        result.Add(new PSTypeName(methodInfo2.MakeGenericMethod(resolvedTypeArguments).ReturnType));
    }
    catch // If we can't build the generic method then just skip it to retain other completion results.
    {
        return;
    }
}
```

I'm evaluating and casting the same thing twice, and I need a new variable. For what? Less indentation? It doesn't make any sense to me.